finetune/base.py

```diff
@@ -837,42 +887,39 @@ def finetune_grid_search_cv(
         return max(aggregated_results, key=lambda x: x[1])[0]
 
     def process_long_sequence(self, zipped_data):
-        arr_encoded = [
-            self.input_pipeline._text_to_ids(d["X"]) for d in zipped_data
-        ]
-
-        flat_array_encoded = []
-        sequence_id = []
-        for i, ae in enumerate(arr_encoded):
-            for sample in ae:
-                flat_array_encoded.append(sample)
-                sequence_id.append(i)
-
         labels, batch_probas = [], []
-        for pred in self._inference(
-                zipped_data,
-                predict_keys=[PredictMode.PROBAS, PredictMode.NORMAL],
-                chunked_length=len(flat_array_encoded),
+        for chunk_idx, (
+            pred,
+            (sequence_id, arr_enc, start_of_doc, end_of_doc),
+        ) in enumerate(
+            zip(
+                self._inference(
+                    zipped_data,
+                    predict_keys=[PredictMode.PROBAS, PredictMode.NORMAL],
+                    chunked_length=0,
+                    list_output=False,
+                ),
+                (
+                    (i, sample, is_start, is_end)
+                    for i, ae in enumerate(
+                        (
+                            self.input_pipeline._text_to_ids(d["X"])
+                            for d in zipped_data
+                        )
+                    )
+                    for sample, is_start, is_end in start_end_gen(ae)
+                ),
+            )
         ):
             normal_pred = pred[PredictMode.NORMAL]
-            if not hasattr(self, 'multi_label'):
+            if not hasattr(self, "multi_label"):
                 normal_pred = np.expand_dims(normal_pred, 0)
-            labels.append(self.input_pipeline.label_encoder.inverse_transform(normal_pred))
-            batch_probas.append(pred[PredictMode.PROBAS])
-
-        if not batch_probas:
-            batch_probas = [None]*len(labels)
-
-        for chunk_idx, (label_seq, proba_seq) in enumerate(zip(labels, batch_probas)):
-            token_end_idx = flat_array_encoded[chunk_idx].token_ends
-            token_start_idx = flat_array_encoded[chunk_idx].token_starts
-            start_of_doc = chunk_idx == 0 or sequence_id[chunk_idx - 1] != sequence_id[chunk_idx]
-            end_of_doc = (
-                chunk_idx + 1 == len(flat_array_encoded) or
-                sequence_id[chunk_idx] != sequence_id[chunk_idx + 1]
-            )
-            useful_start = flat_array_encoded[chunk_idx].useful_start
-            useful_end = flat_array_encoded[chunk_idx].useful_end
+            label_seq = self.input_pipeline.label_encoder.inverse_transform(normal_pred)
+            proba_seq = pred[PredictMode.PROBAS]
+            token_end_idx = arr_enc.token_ends
+            token_start_idx = arr_enc.token_starts
+            useful_start = arr_enc.useful_start
```

Comment:
What are the useful_start and useful_end params? It looks like they are used when we chunk long sequences, but I don't fully understand how

Reply:
We use overlapping chunks but discard any predictions from outside the central window. Unless it is the first or last chunk in the document. These indexes record which portion of each sequence is to be retained.